psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1785 @@
+    /* Allocation for retaining another block of doubly contracted integrals
+       (Q|mi) = C_{mi}(A|mn)[J^{-1/2}]_{AQ} */
+    std::unique_ptr<double[]> QMI_OLD(new double[biggest_block_ / nbf_ * nocc]);
+    /* Allocation for holding terms for constructing the coulomb matrix */
+    std::unique_ptr<double[]> V(new double[naux_]);
+
+    /* Pointer to uncontracted eri terms (A|mn) */
+    double* amn = AMN.get();
+    /* Pointer to contracted eri terms (A|mi) = C_{ n i}(A|m n) */
+    double* ami = AMI.get();
+    /* Pointer to one block of doubly contracted integrals
+       (Q|mi) = C_{mi}(A|mn)[J^{-1/2}]_{AQ} */
+    double* qmi_new = QMI_NEW.get();
+    /* Pointer to another block of doubly contracted integrals
+       (Q|mi) = C_{mi}(A|mn)[J^{-1/2}]_{AQ} */
+    double* qmi_old = QMI_OLD.get();
+    /* Pointer used for swapping addresses */
+    double* qmi_hold;
+    double* v = V.get();
+
+    /* containers for temporary storage of data shaped for sparsity */
+    std::unique_ptr<double[]> P_C(new double[nbf_ * nocc]);
+    std::unique_ptr<double[]> P_D(new double[nbf_]);
+    std::unique_ptr<double[]> P_J(new double[nbf_]);
+
+    /* pointers to temperary storage for data shaped for sparsity */
+    double* pruned_c = P_C.get();
+    double* pruned_d = P_D.get();
+    double* pruned_j = P_J.get();
+
+    /* zeroing out terms that will be accumulated into */
+    memset(j, 0, nbf_ * nbf_ * 8);
+    memset(k, 0, nbf_ * nbf_ * 8);
+    memset(v, 0, naux_ * 8);
+    char first_char = (num_blocks_ == 1 ? 'B' : 'V');
+
+    /* variables to track which block of eri's each pointer holds*/
+    size_t qmi_old_ind = 0U;
+    size_t qmi_new_ind;
+
+    outfile->Printf("num_blocks_ is %zu\n", num_blocks_);
+
+    /* See comments below regarding X_Block */
+    QMI_Block_sparse(first_char, true, 0UL, pruned_c, pruned_d, amn, qmi_old, ami, v, pruned_j, eri);
+    timer_on("DDF pQq big K DGEMM");
+    C_DGEMM('N', 'T', Block_funcs_[0], Block_funcs_[0], naux_ * nocc, 1.0, qmi_old, naux_ * nocc, qmi_old, naux_ * nocc, 1.0, k,
+            nbf_);
+    timer_off("DDF pQq big K DGEMM");
+    /* Loop over blocks of exchange matrix construction. The blocking scheme 
+       is over blocks (A| [m]_i n) where [m]_i is the i^th block over the 
+       primary basis set. The exchange matrix will be constructed as 
+       
+       old_iter = 0
+       x o o o  before loop
+       x x o o  new_iter = 1
+       x o x o  new_iter = 2
+       x o o x  new_iter = 3
+
+       old_iter = 1          old_iter = 2
+       x o o o               x o o o
+       x x o o               x x o o
+       x o x o               x x x o
+       x x x x               x x x x
+         1 2<----new_iter      1<------new_iter
+
+       final: copy lower to upper
+       Legend x: constructed o: empty
+
+       Details can be found by requesting
+       Joseph Senan O'Brien thesis from the Georgia Tech Library
+       In that document (Q|mi) <--> X^Q_{\mu a}
```

Comment:
Is the thesis digitally available? If so please add DOI (or whatever link is available)

Reply:
I'll change to comment to add the URL. It wasn't available when I added that comment